### PR TITLE
[Sort benchmark] Two-stage reduce

### PR DESCRIPTION
**Pull request by frank-lsf**
## Why are these changes needed?

This PR implements the two-stage reduce described in https://docs.google.com/document/d/1G0Fm7RGAxzvCtWylJJvtQHAPKXGJZSfkKBjgKcj0Z2k/edit?usp=sharing

Unlike in the previous version, the number of intermediate partitions now scales linearly with data size, not quadratically, making it feasible to scale to 100TB.

Benchmark result: 1TB sort on 16 nodes takes ~820 seconds. This is slower than before (600-700 seconds), but expected due to unnecessary memory pressure and spilling. The next PR will fix this.

TODO:
- Need back pressure in scheduling to avoid excessive object spilling (in theory, we won't need to spill anything)
- Test scaling to 100TB

cc @stephanie-wang 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by stephanie-wang** [comment]
As discussed offline, I'll review again once some documentation is added.

File: python/ray/experimental/raysort/main.py [hunk 1]
```diff
@@ -30,25 +34,43 @@ def get_args():
...
-        default=4,
+        default=128,
         type=int,
         help="number of map tasks",
     )
     parser.add_argument(
         "--num_reducers",
+        default=16,
+        type=int,
+        help="number of reduce actors",
```
Comment (stephanie-wang) [inline python/ray/experimental/raysort/main.py:51]: Can you clarify that this is the number of second-stage reducers?


It would be helpful to see a description of the design in terms of rounds. Something like this:

M = total # of map tasks
R = total # of rounds
N = total # of nodes

round = 1 set of map tasks which will get merged together. The second-stage reducers merge the results of each round.

File: python/ray/experimental/raysort/main.py [hunk 2]
```diff
@@ -30,25 +34,43 @@ def get_args():
...
+    )
+    parser.add_argument(
```
Comment (stephanie-wang) [inline python/ray/experimental/raysort/main.py:53]: I think it would be more intuitive to say this as maximum number of rounds in flight.

Ideally, I think we always want this to be 2, so that we are producing map tasks for the next round while shuffling the current round. (and setting M, R, and N appropriately to saturate memory).

File: python/ray/experimental/raysort/main.py [hunk 3]
```diff
@@ -30,25 +34,43 @@ def get_args():
...
+        "--merge_concurrency",
         default=4,
         type=int,
-        help="number of reduce tasks",
+        help="max number of merge tasks per worker node",
+    )
+    parser.add_argument(
```
Comment (stephanie-wang) [inline python/ray/experimental/raysort/main.py:59]: Update this as well once we switch to using the round terminology.

File: python/ray/experimental/raysort/main.py [hunk 4]
```diff
@@ -30,25 +34,43 @@ def get_args():
...
+        "--merge_factor",
+        default=16,
+        type=int,
+        help="number of mapper results to buffer before merging",
+    )
+    parser.add_argument(
+        "--reducer_input_chunk",
+        default=100 * 1024 * 1024,
+        type=ByteCount,
+        help="number of bytes to read in memory from each mapper block",
```
Comment (stephanie-wang) [inline python/ray/experimental/raysort/main.py:69]: Clarify that this is for the final-stage reduce, to reduce total memory usage at the reducer.

File: python/ray/experimental/raysort/main.py [hunk 5]
```diff
@@ -190,146 +206,272 @@ def _dummy_sort_and_partition(part: np.ndarray,
...
 
+@tracing_utils.timeit("sort", report_time=True)
+def sort_main(args: Args):
+    parts = _load_manifest(args, constants.INPUT_MANIFEST_FILE)
+    boundaries = sortlib.get_boundaries(args.num_reducers)
+    pgs = _get_placement_groups(args)
+
+    opt = {
+        "num_returns": args.num_reducers,
+        "memory": args.input_part_size * 2,
```
Comment (stephanie-wang) [inline python/ray/experimental/raysort/main.py:348]: Not needed.

File: python/ray/experimental/raysort/main.py [hunk 6]
```diff
@@ -190,146 +206,272 @@ def _dummy_sort_and_partition(part: np.ndarray,
...
+                [f for f in merge_results.flatten() if f is not None],
+                num_returns=num_extra_tasks)
+        merge_results[merge_count, :] = [
+            merge_mapper_blocks.options(placement_group=pgs[r]).remote(
+                args, r, merge_count, *mapper_results[:, r].tolist())
+            for r in range(args.num_reducers)
+        ]
+        merge_count += 1
+
+    for part_id, node, path in parts:
```
Comment (stephanie-wang) [inline python/ray/experimental/raysort/main.py:372]: Might be cleaner if we make this a for loop where the outer loop is per round.

File: python/ray/experimental/raysort/main.py [hunk 7]
```diff
@@ -190,146 +206,272 @@ def _dummy_sort_and_partition(part: np.ndarray,
...
+        "memory": args.input_part_size * 2,
+    }
+    mapper_results = None
+    merge_results = np.empty(
+        (args.num_merged_mappers, args.num_reducers), dtype=object)
+    merge_concurrency = args.merge_concurrency * args.num_workers
+    merge_count = 0
+
+    def submit_merge_tasks():
+        nonlocal merge_count
```
Comment (stephanie-wang) [inline python/ray/experimental/raysort/main.py:357]: Let's try not to use `nonlocal` or lambdas if we can.

File: python/ray/experimental/raysort/main.py [hunk 8]
```diff
@@ -190,146 +206,272 @@ def _dummy_sort_and_partition(part: np.ndarray,
...
+    parts = _load_manifest(args, constants.INPUT_MANIFEST_FILE)
+    boundaries = sortlib.get_boundaries(args.num_reducers)
+    pgs = _get_placement_groups(args)
+
+    opt = {
+        "num_returns": args.num_reducers,
+        "memory": args.input_part_size * 2,
+    }
+    mapper_results = None
+    merge_results = np.empty(
```
Comment (stephanie-wang) [inline python/ray/experimental/raysort/main.py:351]: Add comment to explain what this will contain.

File: python/ray/experimental/raysort/main.py [hunk 9]
```diff
@@ -190,146 +206,272 @@ def _dummy_sort_and_partition(part: np.ndarray,
...
+    ]
+    ray.get([pg.ready() for pg in pgs])
+    return pgs
+
 
+@tracing_utils.timeit("sort", report_time=True)
+def sort_main(args: Args):
+    parts = _load_manifest(args, constants.INPUT_MANIFEST_FILE)
+    boundaries = sortlib.get_boundaries(args.num_reducers)
+    pgs = _get_placement_groups(args)
```
Comment (stephanie-wang) [inline python/ray/experimental/raysort/main.py:344]: Explain what the placement groups are used for and what resources are acquired.

File: python/ray/experimental/raysort/main.py (hunk 2 above)
Comment (stephanie-wang) [reply python/ray/experimental/raysort/main.py:53]: Let's try benchmarking with num_concurrent_rounds=1 and 2.

So 1 should be very conservative and make sure that all of the previous map results have been evicted before the next round is submitted, and we should expect to see no spilling at all.

With 2, we might see spilling if the next round's map tasks are scheduled too soon, but it would be good to measure the difference with this benchmark.

**Review by stephanie-wang — CHANGES_REQUESTED** [review]

File: python/ray/experimental/raysort/main.py [hunk 10]
```diff
@@ -190,146 +208,270 @@ def _dummy_sort_and_partition(part: np.ndarray,
...
+            mapper_results[m, :] = mapper.options(**mapper_opt).remote(
+                args, boundaries, path)
+            part_id += 1
+
+        # Submit merge tasks.
+        merge_results[round, :] = [
+            merge_mapper_blocks.options(placement_group=pgs[r]).remote(
+                args, r, round, *mapper_results[:, r].tolist())
+            for r in range(args.num_reducers)
+        ]
```
Comment (stephanie-wang) [inline python/ray/experimental/raysort/main.py:380]: Probably a good idea to clear the local references to mapper_results here, too, before we do the `ray.wait`.

File: python/ray/experimental/raysort/sortlib.py [hunk 11]
```diff
@@ -21,7 +21,9 @@ def sort_and_partition(part: np.ndarray,
     return blocks
 
 
-def merge_partitions(blocks: List[np.ndarray],
-                     _n: int) -> Iterable[memoryview]:
+def merge_partitions(
+        num_blocks: int, _n: int,
+        get_block: Callable[[int, int], np.ndarray]) -> Iterable[memoryview]:
+    blocks = [get_block(i, 0) for i in range(num_blocks)]
```
Comment (stephanie-wang) [inline python/ray/experimental/raysort/sortlib.py:27]: Is this doing an actual merge? It looks like it's just returning the blocks again?

File: python/ray/experimental/raysort/main.py [hunk 12]
```diff
@@ -190,146 +208,270 @@ def _dummy_sort_and_partition(part: np.ndarray,
...
+    return [part[offset:offset + size] for offset, size in blocks]
 
 
-def _dummy_merge(blocks: List[np.ndarray], _n: int) -> Iterable[memoryview]:
-    for block in blocks:
+def _dummy_merge(
+        num_blocks: int, _n: int,
+        get_block: Callable[[int, int], np.ndarray]) -> Iterable[np.ndarray]:
+    blocks = [((i, 0), get_block(i, 0)) for i in range(num_blocks)]
+    while len(blocks) > 0:
```
Comment (stephanie-wang) [inline python/ray/experimental/raysort/main.py:231]: Can you explain what the dummy_merge does? I found this a bit hard to follow.

File: python/ray/experimental/raysort/main.py [hunk 13]
```diff
@@ -190,146 +208,270 @@ def _dummy_sort_and_partition(part: np.ndarray,
...
+
+    if skip_output:
         for datachunk in merger:
             del datachunk
-        logging.info(f"{task_id} done")
-        return None
     else:
-        pinfo = _make_partition_info(reducer_id, "output")
-        with open(pinfo.path, "wb") as fout:
+        if not remote_writer:
```
Comment (stephanie-wang) [inline python/ray/experimental/raysort/main.py:260]: What's the purpose of the remote writer? Shouldn't we always be writing locally?

File: python/ray/experimental/raysort/main.py [hunk 14]
```diff
@@ -190,146 +208,270 @@ def _dummy_sort_and_partition(part: np.ndarray,
...
+
+    @ray.remote(num_cpus=0, resources={"worker": 1e-3})
+    def _load_block_chunk(pinfo: PartInfo, d: int) -> np.ndarray:
+        return np.fromfile(
+            pinfo.path,
+            dtype=np.uint8,
+            count=args.reducer_input_chunk,
+            offset=d * args.reducer_input_chunk)
+
+    block_chunks = [
```
Comment (stephanie-wang) [inline python/ray/experimental/raysort/main.py:308]: Since this initial prototype is already bypassing the object store for the reduce step, I think it would be simpler here to just load the data into heap memory directly, even if it has to happen serially.

Later, we can measure the difference between this and using an object store-native approach for storing intermediate results in the reduce stage.

File: python/ray/experimental/raysort/main.py [hunk 15]
```diff
@@ -30,27 +35,47 @@ def get_args():
...
+        help="number of second-stage reduce tasks",
+    )
+    parser.add_argument(
+        "--num_concurrent_rounds",
         default=4,
         type=int,
-        help="number of reduce tasks",
+        help="max number of rounds of map/merge tasks in flight",
+    )
+    parser.add_argument(
```
Comment (stephanie-wang) [inline python/ray/experimental/raysort/main.py:66]: Correct me if I'm wrong, but it looks like this and reducer_output_chunk are not actually implemented yet, is that right? How about we remove these for now?

**Review by stephanie-wang — COMMENTED** [review]
The main loop looks a lot cleaner! I left some comments on the lower-level logic.

File: python/ray/experimental/raysort/sortlib.py (hunk 11 above)
Comment (franklsf95) [reply python/ray/experimental/raysort/sortlib.py:27]: No, this is a "fake" implementation. The actual merge is implemented in C++ which needs a build step. Should I commit it? (maybe in another PR)

File: python/ray/experimental/raysort/main.py (hunk 12 above)
Comment (franklsf95) [reply python/ray/experimental/raysort/main.py:231]: It simulates merging `n` blocks. `get_block(i, d)` returns the `d`-th chunk of block `i`. This function repeatedly calls `get_block` until it exhausts all the chunks of all the blocks.

File: python/ray/experimental/raysort/main.py (hunk 13 above)
Comment (franklsf95) [reply python/ray/experimental/raysort/main.py:260]: "remote" as in a Ray remote function. Using placement groups it will always run on the same node. This is for pipelining - offloading the write to disk to another process.

We can do this serially now and it won't have much difference. In the real benchmark you need to write to EBS which would be a bit slower.

File: python/ray/experimental/raysort/main.py (hunk 14 above)
Comment (franklsf95) [reply python/ray/experimental/raysort/main.py:308]: Same as above.

File: python/ray/experimental/raysort/main.py (hunk 15 above)
Comment (franklsf95) [reply python/ray/experimental/raysort/main.py:66]: Yea, it's only implemented in the C++ merge. I can remove these for now.

File: python/ray/experimental/raysort/sortlib.py (hunk 11 above)
Comment (stephanie-wang) [reply python/ray/experimental/raysort/sortlib.py:27]: Ah I see, I got confused because there was also the `dummy_merge` function. Could we replace the `dummy_merge` implementation with this one (this one looks simpler) and throw a NotImplementedError when the user passes the flag for an actual merge?

File: python/ray/experimental/raysort/main.py (hunk 13 above)
Comment (stephanie-wang) [reply python/ray/experimental/raysort/main.py:260]: I see, let's just write serially for now, and we can measure the potential gain for this later. I think it will be easier to understand and measure the performance if we write serially.

File: python/ray/experimental/raysort/main.py (hunk 15 above)
Comment (stephanie-wang) [reply python/ray/experimental/raysort/main.py:66]: Bump?

File: python/ray/experimental/raysort/main.py [hunk 16]
```diff
@@ -190,146 +212,246 @@ def _dummy_sort_and_partition(part: np.ndarray,
...
-        logging.info(f"{task_id} done")
-        return pinfo
+    return pinfo
 
 
-@tracing_utils.timeit("sorting")
-def sort_main():
-    partitions = _load_manifest(constants.INPUT_MANIFEST_FILE)
+@ray.remote(
+    num_cpus=0, resources={"worker": 1e-3})  # TODO: test setting this to 1
```
Comment (stephanie-wang) [inline python/ray/experimental/raysort/main.py:259]: How about setting "worker" to 1 now?

Can you also add a note to explain why we use num_cpus=0?

File: python/ray/experimental/raysort/main.py [hunk 17]
```diff
@@ -190,146 +212,246 @@ def _dummy_sort_and_partition(part: np.ndarray,
...
+
+    def get_block(i, d):
+        if i >= M or d > 0:
+            return None
+        return blocks[i]
+
+    return _merge_impl(args, M, pinfo, get_block)
+
+
+@ray.remote(num_cpus=0, resources={"worker": 1e-3})
```
Comment (stephanie-wang) [inline python/ray/experimental/raysort/main.py:275]: Same here.

File: python/ray/experimental/raysort/main.py [hunk 18]
```diff
@@ -190,146 +212,246 @@ def _dummy_sort_and_partition(part: np.ndarray,
...
-        }
-        blocks = mapper_results[:, r].tolist()
-        ret = reducer.options(**opt).remote(r, *blocks)
-        reducer_results.append(ret)
-
-    reducer_results = ray.get(reducer_results)
+
+    mapper_opt = {
+        "num_returns": args.num_reducers,
+        "num_cpus": 2,  # concurrency = 2
```
Comment (stephanie-wang) [inline python/ray/experimental/raysort/main.py:330]: Why 2 CPUs instead of 1?

**Review by stephanie-wang — APPROVED** [review]
Looks good, just a few more comments to simplify further, but I think we can merge this soon.

**Comment by ericl** [comment]
@stephanie-wang @franklsf95 this is breaking master Windows build, can you revert or disable the test?

**Comment by franklsf95** [comment]
Will disable the test for now.
